File: src/python/batch/models/batch_control.py
"""
Batch Control Record Model

Corresponds to COBOL copybooks: BCHCTL.cpy and BCHCON.cpy
Defines the structure for batch job control and process sequencing.
"""

from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import List, Optional


class BatchStatus(Enum):
    """Batch status codes matching COBOL 88-level conditions in BCHCTL.cpy."""
    READY = "R"
    ACTIVE = "A"
    WAITING = "W"
    DONE = "D"
    ERROR = "E"


class ReturnCode(Enum):
    """Return code thresholds matching COBOL constants in BCHCON.cpy."""
    SUCCESS = 0
    WARNING = 4
    ERROR = 8
    SEVERE = 12
    CRITICAL = 16


class ProcessType(Enum):
    """Process type codes matching COBOL constants in BCHCON.cpy."""
    INITIAL = "INI"
    UPDATE = "UPD"
    REPORT = "RPT"
    CLEANUP = "CLN"


class DependencyType(Enum):
    """Dependency type codes matching COBOL constants in BCHCON.cpy."""
    REQUIRED = "R"
    OPTIONAL = "O"
    EXCLUSIVE = "X"


@dataclass
class PrerequisiteJob:
    """
    Prerequisite job definition.
    
    Corresponds to BCT-PREREQ-JOBS in BCHCTL.cpy:
    - BCT-PREREQ-NAME: PIC X(8)
    - BCT-PREREQ-SEQ: PIC 9(4)
    - BCT-PREREQ-RC: PIC S9(4) COMP
    """
    name: str
    sequence: int = 0
    max_return_code: int = 0

    def __post_init__(self) -> None:
        self.name = str(self.name).ljust(8)[:8]


@dataclass
class BatchControlKey:
    """
    Batch control key structure.
    
    Corresponds to BCT-KEY in BCHCTL.cpy:
    - BCT-JOB-NAME: PIC X(8)
    - BCT-PROCESS-DATE: PIC X(8)
    - BCT-SEQUENCE-NO: PIC 9(4)
    """
    job_name: str
    process_date: str  # YYYYMMDD format
    sequence_no: int = 0

    def __post_init__(self) -> None:
        self.job_name = str(self.job_name).ljust(8)[:8]
        self.process_date = str(self.process_date).ljust(8)[:8]

    def to_string(self) -> str:
        """Convert key to string for comparison and storage."""
        return f"{self.job_name}{self.process_date}{self.sequence_no:04d}"

    @classmethod
    def from_string(cls, key_string: str) -> "BatchControlKey":
        """Parse key from string representation."""
        return cls(
            job_name=key_string[0:8],
            process_date=key_string[8:16],
            sequence_no=int(key_string[16:20]),
        )


@dataclass
class ProcessControl:
    """
    Process control structure.
    
    Corresponds to BCT-PROCESS-CONTROL in BCHCTL.cpy:
    - BCT-STEP-NAME: PIC X(8)
    - BCT-PROGRAM-NAME: PIC X(8)
    - BCT-START-TIME: PIC X(8)
    - BCT-END-TIME: PIC X(8)
    """
    step_name: str = ""
    program_name: str = ""
    start_time: str = ""
    end_time: str = ""

    def __post_init__(self) -> None:
        self.step_name = str(self.step_name).ljust(8)[:8]
        self.program_name = str(self.program_name).ljust(8)[:8]
        self.start_time = str(self.start_time).ljust(8)[:8]
        self.end_time = str(self.end_time).ljust(8)[:8]


@dataclass
class Dependencies:
    """
    Dependencies structure.
    
    Corresponds to BCT-DEPENDENCIES in BCHCTL.cpy:
    - BCT-PREREQ-COUNT: PIC 9(2) COMP
    - BCT-PREREQ-JOBS: OCCURS 10 TIMES
    """
    prereq_count: int = 0
    prereq_jobs: List[PrerequisiteJob] = field(default_factory=list)

    def add_prerequisite(
        self, name: str, sequence: int = 0, max_return_code: int = 0
    ) -> None:
        """Add a prerequisite job."""
        if self.prereq_count >= 10:
            raise ValueError("Maximum 10 prerequisites allowed")
        self.prereq_jobs.append(
            PrerequisiteJob(name=name, sequence=sequence, max_return_code=max_return_code)
        )
        self.prereq_count += 1


@dataclass
class ReturnInfo:
    """
    Return information structure.
    
    Corresponds to BCT-RETURN-INFO in BCHCTL.cpy:
    - BCT-RETURN-CODE: PIC S9(4) COMP
    - BCT-ERROR-DESC: PIC X(80)
    """
    return_code: int = 0
    error_desc: str = ""

    def __post_init__(self) -> None:
        self.error_desc = str(self.error_desc).ljust(80)[:80]


@dataclass
class Statistics:
    """
    Statistics structure.
    
    Corresponds to BCT-STATISTICS in BCHCTL.cpy:
    - BCT-RESTART-COUNT: PIC 9(2) COMP
    - BCT-ATTEMPT-TS: PIC X(26)
    - BCT-COMPLETE-TS: PIC X(26)
    """
    restart_count: int = 0
    attempt_ts: str = ""
    complete_ts: str = ""

    def __post_init__(self) -> None:
        self.attempt_ts = str(self.attempt_ts).ljust(26)[:26]
        self.complete_ts = str(self.complete_ts).ljust(26)[:26]


@dataclass
class BatchControlRecord:
    """
    Complete batch control record structure.
    
    Corresponds to BATCH-CONTROL-RECORD in BCHCTL.cpy.
    """
    key: BatchControlKey
    status: BatchStatus = BatchStatus.READY
    process_control: ProcessControl = field(default_factory=ProcessControl)
    dependencies: Dependencies = field(default_factory=Dependencies)
    return_info: ReturnInfo = field(default_factory=ReturnInfo)
    statistics: Statistics = field(default_factory=Statistics)
    filler: str = ""

    def __post_init__(self) -> None:
        self.filler = " " * 50
        if isinstance(self.status, str):
            self.status = BatchStatus(self.status)

    @property
    def job_name(self) -> str:
        return self.key.job_name

    @property
    def process_date(self) -> str:
        return self.key.process_date

    @property
    def sequence_no(self) -> int:
        return self.key.sequence_no

    @property
    def return_code(self) -> int:
        return self.return_info.return_code

    @return_code.setter
    def return_code(self, value: int) -> None:
        self.return_info.return_code = value

    def is_ready(self) -> bool:
        return self.status == BatchStatus.READY

    def is_active(self) -> bool:
        return self.status == BatchStatus.ACTIVE

    def is_waiting(self) -> bool:
        return self.status == BatchStatus.WAITING

    def is_done(self) -> bool:
        return self.status == BatchStatus.DONE

    def is_error(self) -> bool:
        return self.status == BatchStatus.ERROR

    def start_processing(self, program_name: str = "") -> None:
        """Mark the job as active and record start time."""
        self.status = BatchStatus.ACTIVE
        self.process_control.program_name = program_name
        self.process_control.start_time = datetime.now().strftime("%H%M%S%f")[:8]
        self.statistics.attempt_ts = datetime.now().isoformat()

    def complete_processing(self, return_code: int = 0, error_desc: str = "") -> None:
        """Mark the job as complete and record end time."""
        self.status = BatchStatus.DONE if return_code <= 4 else BatchStatus.ERROR
        self.process_control.end_time = datetime.now().strftime("%H%M%S%f")[:8]
        self.return_info.return_code = return_code
        self.return_info.error_desc = error_desc
        self.statistics.complete_ts = datetime.now().isoformat()

    def increment_restart(self) -> bool:
        """
        Increment restart count and check if max restarts exceeded.
        Returns True if restart is allowed, False otherwise.
        """
        self.statistics.restart_count += 1
        return self.statistics.restart_count <= 3

    def check_prerequisites(
        self, completed_jobs: dict[str, int]
    ) -> tuple[bool, str]:
        """
        Check if all prerequisites are satisfied.
        
        Args:
            completed_jobs: Dict mapping job names to their return codes
            
        Returns:
            Tuple of (satisfied, message)
        """
        for prereq in self.dependencies.prereq_jobs:
            prereq_name = prereq.name.strip()
            if prereq_name not in completed_jobs:
                return False, f"Prerequisite {prereq_name} not completed"
            if completed_jobs[prereq_name] > prereq.max_return_code:
                return (
                    False,
                    f"Prerequisite {prereq_name} return code "
                    f"{completed_jobs[prereq_name]} exceeds max {prereq.max_return_code}",
                )
        return True, "All prerequisites satisfied"

    def to_dict(self) -> dict:
        """Convert record to dictionary for serialization."""
        return {
            "key": {
                "job_name": self.key.job_name,
                "process_date": self.key.process_date,
                "sequence_no": self.key.sequence_no,
            },
            "status": self.status.value,
            "process_control": {
                "step_name": self.process_control.step_name,
                "program_name": self.process_control.program_name,
                "start_time": self.process_control.start_time,
                "end_time": self.process_control.end_time,
            },
            "dependencies": {
                "prereq_count": self.dependencies.prereq_count,
                "prereq_jobs": [
                    {
                        "name": p.name,
                        "sequence": p.sequence,
                        "max_return_code": p.max_return_code,
                    }
                    for p in self.dependencies.prereq_jobs
                ],
            },
            "return_info": {
                "return_code": self.return_info.return_code,
                "error_desc": self.return_info.error_desc,
            },
            "statistics": {
                "restart_count": self.statistics.restart_count,
                "attempt_ts": self.statistics.attempt_ts,
                "complete_ts": self.statistics.complete_ts,
            },
        }

    @classmethod
    def from_dict(cls, data: dict) -> "BatchControlRecord":
        """Create record from dictionary."""
        deps = Dependencies(prereq_count=data["dependencies"]["prereq_count"])
        for p in data["dependencies"]["prereq_jobs"]:
            deps.prereq_jobs.append(
                PrerequisiteJob(
                    name=p["name"],
                    sequence=p["sequence"],
                    max_return_code=p["max_return_code"],
                )
            )
        
        return cls(
            key=BatchControlKey(
                job_name=data["key"]["job_name"],
                process_date=data["key"]["process_date"],
                sequence_no=data["key"]["sequence_no"],
            ),
            status=BatchStatus(data["status"]),
            process_control=ProcessControl(
                step_name=data["process_control"]["step_name"],
                program_name=data["process_control"]["program_name"],
                start_time=data["process_control"]["start_time"],
                end_time=data["process_control"]["end_time"],
            ),
            dependencies=deps,
            return_info=ReturnInfo(
                return_code=data["return_info"]["return_code"],
                error_desc=data["return_info"]["error_desc"],
            ),
            statistics=Statistics(
                restart_count=data["statistics"]["restart_count"],
                attempt_ts=data["statistics"]["attempt_ts"],
                complete_ts=data["statistics"]["complete_ts"],
            ),
        )

File: rust-backend/crates/batch/src/batch_control.rs
//! Batch control processor.
//!
//! Ported from COBOL program `BCHCTL00.cbl` and copybooks `BCHCTL.cpy` /
//! `BCHCON.cpy`.
//!
//! Manages batch-job lifecycle: initialisation, prerequisite checking,
//! status updates, checkpoint/commit, and termination.

use std::collections::HashMap;
use std::fmt;

use chrono::{NaiveDate, Utc};
use serde::{Deserialize, Serialize};

// ---------------------------------------------------------------------------
// BCHCON.cpy — Batch control constants
// ---------------------------------------------------------------------------

/// Process status values (mirrors BCT-STAT-VALUES).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum JobStatus {
    Ready,
    Active,
    Waiting,
    Done,
    Error,
}

impl JobStatus {
    pub fn code(&self) -> char {
        match self {
            Self::Ready => 'R',
            Self::Active => 'A',
            Self::Waiting => 'W',
            Self::Done => 'D',
            Self::Error => 'E',
        }
    }

    pub fn from_code(c: char) -> Option<Self> {
        match c {
            'R' => Some(Self::Ready),
            'A' => Some(Self::Active),
            'W' => Some(Self::Waiting),
            'D' => Some(Self::Done),
            'E' => Some(Self::Error),
            _ => None,
        }
    }

    pub fn is_terminal(self) -> bool {
        matches!(self, Self::Done | Self::Error)
    }
}

impl fmt::Display for JobStatus {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.code())
    }
}

/// Return-code thresholds (mirrors BCT-RC-THRESHOLDS).
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub struct ReturnCode(pub i16);

impl ReturnCode {
    pub const SUCCESS: Self = Self(0);
    pub const WARNING: Self = Self(4);
    pub const ERROR: Self = Self(8);
    pub const SEVERE: Self = Self(12);
    pub const CRITICAL: Self = Self(16);
}

impl fmt::Display for ReturnCode {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

/// Process type (mirrors BCT-PROC-TYPES).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ProcessType {
    Initial,
    Update,
    Report,
    Cleanup,
}

impl ProcessType {
    pub fn code(&self) -> &'static str {
        match self {
            Self::Initial => "INI",
            Self::Update => "UPD",
            Self::Report => "RPT",
            Self::Cleanup => "CLN",
        }
    }

    pub fn from_code(code: &str) -> Option<Self> {
        match code {
            "INI" => Some(Self::Initial),
            "UPD" => Some(Self::Update),
            "RPT" => Some(Self::Report),
            "CLN" => Some(Self::Cleanup),
            _ => None,
        }
    }
}

/// Dependency type (mirrors BCT-DEP-TYPES).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum DependencyType {
    Required,
    Optional,
    Exclusive,
}

impl DependencyType {
    pub fn code(&self) -> char {
        match self {
            Self::Required => 'R',
            Self::Optional => 'O',
            Self::Exclusive => 'X',
        }
    }
}

// ---------------------------------------------------------------------------
// BCHCTL.cpy — Batch control record
// ---------------------------------------------------------------------------

/// A prerequisite entry for a batch job (mirrors BCT-PREREQ-JOBS).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Prerequisite {
    pub job_name: String,
    pub sequence_no: u16,
    pub max_return_code: i16,
}

/// Batch control record (mirrors BATCH-CONTROL-RECORD).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BatchControlRecord {
    // -- BCT-KEY --
    pub job_name: String,
    pub process_date: NaiveDate,
    pub sequence_no: u16,
    // -- BCT-DATA --
    pub status: JobStatus,
    pub step_name: String,
    pub program_name: String,
    pub start_time: Option<String>,
    pub end_time: Option<String>,
    // -- BCT-DEPENDENCIES --
    pub prerequisites: Vec<Prerequisite>,
    // -- BCT-RETURN-INFO --
    pub return_code: i16,
    pub error_desc: String,
    // -- BCT-STATISTICS --
    pub restart_count: u16,
    pub attempt_ts: Option<String>,
    pub complete_ts: Option<String>,
}

impl BatchControlRecord {
    pub fn new(job_name: impl Into<String>, process_date: NaiveDate, sequence_no: u16) -> Self {
        Self {
            job_name: job_name.into(),
            process_date,
            sequence_no,
            status: JobStatus::Ready,
            step_name: String::new(),
            program_name: String::new(),
            start_time: None,
            end_time: None,
            prerequisites: Vec::new(),
            return_code: 0,
            error_desc: String::new(),
            restart_count: 0,
            attempt_ts: None,
            complete_ts: None,
        }
    }
}

// ---------------------------------------------------------------------------
// BatchController — port of BCHCTL00
// ---------------------------------------------------------------------------

/// Errors produced by [`BatchController`].
#[derive(Debug, Clone, thiserror::Error)]
pub enum BatchControlError {
    #[error("job not found: {0}")]
    JobNotFound(String),

    #[error("invalid function: {0}")]
    InvalidFunction(String),

    #[error("prerequisites not satisfied for job {0}")]
    PrerequisitesNotMet(String),

    #[error("invalid state transition: {from:?} -> {to:?} for job {job}")]
    InvalidTransition {
        job: String,
        from: JobStatus,
        to: JobStatus,
    },

    #[error("max restarts ({max}) exceeded for job {job}")]
    MaxRestartsExceeded { job: String, max: u16 },

    #[error("batch control error: {0}")]
    Other(String),
}

/// In-memory batch controller (port of BCHCTL00).
///
/// Manages a table of [`BatchControlRecord`]s keyed by `(job_name, process_date)`.
/// Provides the four COBOL entry-points: INIT, CHECK, UPDATE, TERMINATE.
#[derive(Debug, Clone)]
pub struct BatchController {
    records: HashMap<String, BatchControlRecord>,
    max_restarts: u16,
}

impl Default for BatchController {
    fn default() -> Self {
        Self::new()
    }
}

impl BatchController {
    pub fn new() -> Self {
        Self {
            records: HashMap::new(),
            max_restarts: 3, // BCT-MAX-RESTARTS
        }
    }

    pub fn with_max_restarts(mut self, max: u16) -> Self {
        self.max_restarts = max;
        self
    }

    /// Register a control record (mirrors 1000-PROCESS-INITIALIZE).
    pub fn register_job(&mut self, record: BatchControlRecord) {
        let key = Self::make_key(&record.job_name, &record.process_date);
        self.records.insert(key, record);
    }

    /// Initialise a job — set it to Active with a start timestamp
    /// (mirrors FUNC-INIT → 1000-PROCESS-INITIALIZE).
    pub fn init_job(
        &mut self,
        job_name: &str,
        process_date: &NaiveDate,
    ) -> Result<ReturnCode, BatchControlError> {
        let key = Self::make_key(job_name, process_date);
        let rec = self
            .records
            .get_mut(&key)
            .ok_or_else(|| BatchControlError::JobNotFound(job_name.to_string()))?;

        // 1300-VALIDATE-PROCESS: must be Ready or previously failed (for restart).
        match rec.status {
            JobStatus::Ready => {}
            JobStatus::Error => {
                if rec.restart_count >= self.max_restarts {
                    return Err(BatchControlError::MaxRestartsExceeded {
                        job: job_name.to_string(),
                        max: self.max_restarts,
                    });
                }
                rec.restart_count += 1;
            }
            other => {
                return Err(BatchControlError::InvalidTransition {
                    job: job_name.to_string(),
                    from: other,
                    to: JobStatus::Active,
                });
            }
        }

        // 1400-UPDATE-START-STATUS
        rec.status = JobStatus::Active;
        let now = Utc::now();
        rec.start_time = Some(now.format("%H:%M:%S").to_string());
        rec.attempt_ts = Some(now.to_rfc3339());
        rec.return_code = 0;
        rec.error_desc.clear();

        Ok(ReturnCode::SUCCESS)
    }

    /// Check whether all prerequisites for a job are satisfied
    /// (mirrors FUNC-CHEK → 2000-CHECK-PREREQUISITES).
    pub fn check_prerequisites(
        &self,
        job_name: &str,
        process_date: &NaiveDate,
    ) -> Result<ReturnCode, BatchControlError> {
        let key = Self::make_key(job_name, process_date);
        let rec = self
            .records
            .get(&key)
            .ok_or_else(|| BatchControlError::JobNotFound(job_name.to_string()))?;

        // 2200-CHECK-DEPENDENCIES
        for prereq in &rec.prerequisites {
            let dep_key = Self::make_key(&prereq.job_name, process_date);
            match self.records.get(&dep_key) {
                Some(dep_rec) => {
                    if dep_rec.status != JobStatus::Done {
                        return Ok(ReturnCode::WARNING);
                    }
                    if dep_rec.return_code > prereq.max_return_code {
                        return Ok(ReturnCode::ERROR);
                    }
                }
                None => return Ok(ReturnCode::WARNING),
            }
        }

        Ok(ReturnCode::SUCCESS)
    }

    /// Update the status of a running job
    /// (mirrors FUNC-UPDT → 3000-UPDATE-STATUS).
    pub fn update_status(
        &mut self,
        job_name: &str,
        process_date: &NaiveDate,
        new_status: JobStatus,
        return_code: i16,
        error_desc: Option<&str>,
    ) -> Result<ReturnCode, BatchControlError> {
        let key = Self::make_key(job_name, process_date);
        let rec = self
            .records
            .get_mut(&key)
            .ok_or_else(|| BatchControlError::JobNotFound(job_name.to_string()))?;

        // 3200-UPDATE-PROCESS-STATUS
        rec.status = new_status;
        rec.return_code = return_code;
        if let Some(desc) = error_desc {
            rec.error_desc = desc.to_string();
        }

        Ok(ReturnCode::SUCCESS)
    }

    /// Terminate a job — mark as Done/Error and record completion timestamp
    /// (mirrors FUNC-TERM → 4000-PROCESS-TERMINATE).
    pub fn terminate_job(
        &mut self,
        job_name: &str,
        process_date: &NaiveDate,
        return_code: i16,
    ) -> Result<ReturnCode, BatchControlError> {
        let key = Self::make_key(job_name, process_date);
        let rec = self
            .records
            .get_mut(&key)
            .ok_or_else(|| BatchControlError::JobNotFound(job_name.to_string()))?;

        // 4100-UPDATE-COMPLETION
        let now = Utc::now();
        rec.end_time = Some(now.format("%H:%M:%S").to_string());
        rec.complete_ts = Some(now.to_rfc3339());
        rec.return_code = return_code;

        rec.status = if return_code <= ReturnCode::WARNING.0 {
            JobStatus::Done
        } else {
            JobStatus::Error
        };

        Ok(ReturnCode::SUCCESS)
    }

    /// Retrieve a record by key.
    pub fn get_record(
        &self,
        job_name: &str,
        process_date: &NaiveDate,
    ) -> Option<&BatchControlRecord> {
        let key = Self::make_key(job_name, process_date);
        self.records.get(&key)
    }

    /// List all records (snapshot).
    pub fn records(&self) -> impl Iterator<Item = &BatchControlRecord> {
        self.records.values()
    }

    fn make_key(job_name: &str, process_date: &NaiveDate) -> String {
        format!("{job_name}:{process_date}")
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::NaiveDate;

    fn date(y: i32, m: u32, d: u32) -> NaiveDate {
        NaiveDate::from_ymd_opt(y, m, d).unwrap()
    }

    #[test]
    fn job_status_roundtrip() {
        for s in [
            JobStatus::Ready,
            JobStatus::Active,
            JobStatus::Waiting,
            JobStatus::Done,
            JobStatus::Error,
        ] {
            assert_eq!(JobStatus::from_code(s.code()), Some(s));
        }
        assert!(JobStatus::from_code('Z').is_none());
    }

    #[test]
    fn init_sets_active() {
        let mut ctl = BatchController::new();
        let rec = BatchControlRecord::new("TRNVAL00", date(2024, 1, 15), 1);
        ctl.register_job(rec);

        let rc = ctl.init_job("TRNVAL00", &date(2024, 1, 15)).unwrap();
        assert_eq!(rc, ReturnCode::SUCCESS);

        let rec = ctl.get_record("TRNVAL00", &date(2024, 1, 15)).unwrap();
        assert_eq!(rec.status, JobStatus::Active);
        assert!(rec.start_time.is_some());
    }

    #[test]
    fn init_from_error_increments_restart() {
        let mut ctl = BatchController::new();
        let mut rec = BatchControlRecord::new("LOAD01", date(2024, 1, 15), 1);
        rec.status = JobStatus::Error;
        ctl.register_job(rec);

        ctl.init_job("LOAD01", &date(2024, 1, 15)).unwrap();
        let rec = ctl.get_record("LOAD01", &date(2024, 1, 15)).unwrap();
        assert_eq!(rec.restart_count, 1);
        assert_eq!(rec.status, JobStatus::Active);
    }

    #[test]
    fn init_from_active_fails() {
        let mut ctl = BatchController::new();
        let mut rec = BatchControlRecord::new("JOB1", date(2024, 1, 15), 1);
        rec.status = JobStatus::Active;
        ctl.register_job(rec);

        let err = ctl.init_job("JOB1", &date(2024, 1, 15)).unwrap_err();
        assert!(matches!(err, BatchControlError::InvalidTransition { .. }));
    }

    #[test]
    fn max_restarts_exceeded() {
        let mut ctl = BatchController::new().with_max_restarts(2);
        let mut rec = BatchControlRecord::new("JOB2", date(2024, 1, 15), 1);
        rec.status = JobStatus::Error;
        rec.restart_count = 2;
        ctl.register_job(rec);

        let err = ctl.init_job("JOB2", &date(2024, 1, 15)).unwrap_err();
        assert!(matches!(err, BatchControlError::MaxRestartsExceeded { .. }));
    }

    #[test]
    fn prerequisites_all_done() {
        let mut ctl = BatchController::new();
        let d = date(2024, 1, 15);

        let mut dep = BatchControlRecord::new("STEP_A", d, 1);
        dep.status = JobStatus::Done;
        dep.return_code = 0;
        ctl.register_job(dep);

        let mut main_job = BatchControlRecord::new("STEP_B", d, 2);
        main_job.prerequisites.push(Prerequisite {
            job_name: "STEP_A".into(),
            sequence_no: 1,
            max_return_code: 4,
        });
        ctl.register_job(main_job);

        let rc = ctl.check_prerequisites("STEP_B", &d).unwrap();
        assert_eq!(rc, ReturnCode::SUCCESS);
    }

    #[test]
    fn prerequisites_not_done() {
        let mut ctl = BatchController::new();
        let d = date(2024, 1, 15);

        let dep = BatchControlRecord::new("DEP1", d, 1); // still Ready
        ctl.register_job(dep);

        let mut job = BatchControlRecord::new("JOB_X", d, 2);
        job.prerequisites.push(Prerequisite {
            job_name: "DEP1".into(),
            sequence_no: 1,
            max_return_code: 4,
        });
        ctl.register_job(job);

        let rc = ctl.check_prerequisites("JOB_X", &d).unwrap();
        assert_eq!(rc, ReturnCode::WARNING);
    }

    #[test]
    fn prerequisites_bad_return_code() {
        let mut ctl = BatchController::new();
        let d = date(2024, 1, 15);

        let mut dep = BatchControlRecord::new("DEP2", d, 1);
        dep.status = JobStatus::Done;
        dep.return_code = 8;
        ctl.register_job(dep);

        let mut job = BatchControlRecord::new("JOB_Y", d, 2);
        job.prerequisites.push(Prerequisite {
            job_name: "DEP2".into(),
            sequence_no: 1,
            max_return_code: 4,
        });
        ctl.register_job(job);

        let rc = ctl.check_prerequisites("JOB_Y", &d).unwrap();
        assert_eq!(rc, ReturnCode::ERROR);
    }

    #[test]
    fn terminate_success() {
        let mut ctl = BatchController::new();
        let d = date(2024, 1, 15);
        ctl.register_job(BatchControlRecord::new("JOB_T", d, 1));
        ctl.init_job("JOB_T", &d).unwrap();

        let rc = ctl.terminate_job("JOB_T", &d, 0).unwrap();
        assert_eq!(rc, ReturnCode::SUCCESS);

        let rec = ctl.get_record("JOB_T", &d).unwrap();
        assert_eq!(rec.status, JobStatus::Done);
        assert!(rec.complete_ts.is_some());
    }

    #[test]
    fn terminate_with_error() {
        let mut ctl = BatchController::new();
        let d = date(2024, 1, 15);
        ctl.register_job(BatchControlRecord::new("JOB_E", d, 1));
        ctl.init_job("JOB_E", &d).unwrap();

        ctl.terminate_job("JOB_E", &d, 8).unwrap();
        let rec = ctl.get_record("JOB_E", &d).unwrap();
        assert_eq!(rec.status, JobStatus::Error);
    }

    #[test]
    fn update_status_sets_fields() {
        let mut ctl = BatchController::new();
        let d = date(2024, 1, 15);
        ctl.register_job(BatchControlRecord::new("UPD_J", d, 1));
        ctl.init_job("UPD_J", &d).unwrap();

        ctl.update_status("UPD_J", &d, JobStatus::Waiting, 4, Some("waiting on deps"))
            .unwrap();
        let rec = ctl.get_record("UPD_J", &d).unwrap();
        assert_eq!(rec.status, JobStatus::Waiting);
        assert_eq!(rec.return_code, 4);
        assert_eq!(rec.error_desc, "waiting on deps");
    }

    #[test]
    fn return_code_ordering() {
        assert!(ReturnCode::SUCCESS < ReturnCode::WARNING);
        assert!(ReturnCode::WARNING < ReturnCode::ERROR);
        assert!(ReturnCode::ERROR < ReturnCode::SEVERE);
        assert!(ReturnCode::SEVERE < ReturnCode::CRITICAL);
    }

    #[test]
    fn process_type_roundtrip() {
        for pt in [
            ProcessType::Initial,
            ProcessType::Update,
            ProcessType::Report,
            ProcessType::Cleanup,
        ] {
            assert_eq!(ProcessType::from_code(pt.code()), Some(pt));
        }
        assert!(ProcessType::from_code("ZZZ").is_none());
    }

    #[test]
    fn job_not_found() {
        let ctl = BatchController::new();
        let err = ctl
            .check_prerequisites("NOPE", &date(2024, 1, 1))
            .unwrap_err();
        assert!(matches!(err, BatchControlError::JobNotFound(_)));
    }
}
